src/sentry/utils/kafka_config.py

```diff
@@ -0,0 +1,76 @@
+from __future__ import absolute_import
+import six
+from django.conf import settings
+
+SUPPORTED_KAFKA_CONFIGURATION = (
+    # Check https://github.com/edenhill/librdkafka/blob/master/CONFIGURATION.md
+    # for the full list of available options
+    "bootstrap.servers",
+    "compression.type",
+    "message.max.bytes",
+    "sasl.mechanism",
+    "sasl.username",
+    "sasl.password",
+    "security.protocol",
+    "socket.timeout.ms",
+)
+COMMON_SECTION = "common"
+PRODUCERS_SECTION = "producers"
+CONSUMERS_SECTION = "consumers"
+ADMIN_SECTION = "admin"
+KNOWN_SECTIONS = (COMMON_SECTION, PRODUCERS_SECTION, CONSUMERS_SECTION, ADMIN_SECTION)
+
+
+def _get_legacy_kafka_cluster_options(cluster_name):
+    options = settings.KAFKA_CLUSTERS[cluster_name]
+
+    options = {k: v for k, v in options.items() if k not in KNOWN_SECTIONS}
+    if "bootstrap.servers" in options:
+        if isinstance(options["bootstrap.servers"], (list, tuple)):
+            options["bootstrap.servers"] = ",".join(options["bootstrap.servers"])
+    return options
+
+
+def _get_kafka_cluster_options(
+    cluster_name, config_section, with_legacy=False, override_params=None
+):
+    options = {}
+    custom_options = settings.KAFKA_CLUSTERS[cluster_name].get(config_section, {})
+    common_options = settings.KAFKA_CLUSTERS[cluster_name].get(COMMON_SECTION, {})
+    legacy_options = _get_legacy_kafka_cluster_options(cluster_name)
+    if with_legacy and legacy_options:
+        # producer uses all legacy_options
+        options.update(legacy_options)
+    elif "bootstrap.servers" in legacy_options:
+        # legacy override of bootstrap.servers should be preserved
+        options["bootstrap.servers"] = legacy_options["bootstrap.servers"]
```

Comment:
This is going to be very hard to grasp. We need to change how this is structured.
Basically what you are trying to do is to pick all legacy options for producers (if there are legacy options) and, for consumers, pick only bootstrap.servers from legacy options.
Why not rewriting it as
```
#Rename with_legacy into only_bootstrap and invert
if legacy_options:
   if only_bootstrap:
       assert "bootstrap.servers" in legacy_options
       options["bootstrap.servers"] = legacy_options["bootstrap.servers"]
   else :
       options.update(legacy_options)
else:
   # same as before.
```

So it should be clear that picking the bootstrap only from the legacy option is not a way to override other options but it is meant to build legacy options that discard everything that is not legacy.

Reply:
yup, your version is way more readable. this ended up like this due to many rewrittings.

updated